appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockVisibleExtension.java

```diff
@@ -0,0 +1,281 @@
+              }
+
+              return null;
+            });
+  }
+
+  private Promise<String> fetchFileContent(String path) {
+    return Promise.<String>call(
+            "Server error: Could not load file: " + path,
+            cb -> Ode.getInstance().getProjectService().load(projectId, path, cb))
+        .then(Promise::resolve);
+  }
+
+  private void initializeWorker(String mockScriptSrc) {
+    final String[] workerSrc = buildWorkerSource(mockScriptSrc);
+    final BlobOptions blobOpts = BlobOptions.create("text/javascript", "transparent");
+    final Blob blob = new Blob(workerSrc, blobOpts);
+
+    workerUrl = URL.createObjectURL(blob);
+    final WorkerOptions workerOpts = WorkerOptions.create("module");
+    worker = new Worker(workerUrl, workerOpts);
+
+    worker.addEventListener("message", this::handleMessageEvent);
+    worker.addEventListener("error", this::handleErrorEvent);
+  }
+
+  private String[] buildWorkerSource(String mockScriptSrc) {
+    final String escapedScript = JsonUtils.escapeValue(mockScriptSrc);
+    final String baseUrl = Window.Location.getProtocol() + "//" + Window.Location.getHost();
+
+    return new String[] {
+      "import { parseHTML } from '" + baseUrl + "/static/linkedom/linkedom.min.js';\n",
+      "const Mock = {\n",
+      "  document: undefined,\n",
+      "  template: () => undefined,\n",
+      "  onPropertyChange: (property) => undefined,\n",
+      "  onError: (error) => undefined,\n",
+      "};\n",
+      "const mockScript = new Function('Mock', " + escapedScript + ");\n",
+      "mockScript(Mock);\n",
+      "const mockHTML = Mock.template();\n",
+      "postMessage({\n",
+      "  type: '" + MOCK_RENDER_REQUEST + "',\n",
+      "  data: mockHTML\n",
+      "});\n",
+      "Mock.document = parseHTML(mockHTML).document;\n",
+      "onmessage = (msg) => {\n",
+      "  Mock.onPropertyChange(JSON.parse(msg.data));\n",
+      "  postMessage({\n",
+      "    type: '" + MOCK_RENDER_REQUEST + "',\n",
+      "    data: Mock.document.toString()\n",
+      "  });\n",
+      "};\n",
+      "onerror = (err) => Mock.onError(err);\n",
+    };
+  }
+
+  private void handleMessageEvent(MessageEvent<WorkerMessage> msg) {
+    final WorkerMessage msgData = msg.getData();
+    Ode.CLog("worker.message: type: " + msgData.getType());
+
+    if (msgData.getType().equals(MOCK_RENDER_REQUEST)) {
+      final String htmlStr = msgData.getData().toString();
+      Ode.CLog("worker.data: dirty: " + htmlStr);
+      final String sanitizedData = DOMPurify.sanitize(htmlStr);
+      Ode.CLog("worker.data: clean: " + sanitizedData);
+
+      final HTMLPanel htmlPanel = new HTMLPanel(sanitizedData);
+      htmlPanel.setStylePrimaryName("ode-SimpleMockComponent");
+      shadowRoot.removeAllChildren();
+      shadowRoot.appendChild(htmlPanel.getElement());
+    }
+  }
+
+  // TODO
+  private native void handleErrorEvent(Object error) /*-{
+    console.log(error)
```

Comment:
nit: use `console.error`

Reply:
It's supposed to be replaced with better error handling before merging.